# Hash 함수

- 데이터를 hash라는 고정길이 값으로 매핑하는 함수
- 주요 특징
    - 같은 데이터에 대해서 늘 같은 결과 제공
    - 빠른 계산 속도
    - 데이터의 크기와 관계없이 결과는 고정 길이 값.
    - Hash 값으로부터 원래 데이터 복원 불가능.
    

## 정수의 Hash 값.
- 64 비트 시스템의 경우 [$ -2^{61} + 2 , 2^{61} - 2 $] 범위의 정수는 hash 값이 자신과 같음

In [2]:
n = 2**61
print(f"{'n':^24} {'hash(n)':^24}'")
for i in range(-2, 4):
    print(f"{i:2}, {-n+i:20}, {hash(-n+i):20}")
print()
for i in range(-3, 2):
    print(f"{i:2}, {n+i:20}, {hash(n+i):20}")

           n                     hash(n)         '
-2, -2305843009213693954,                   -3
-1, -2305843009213693953,                   -2
 0, -2305843009213693952,                   -2
 1, -2305843009213693951,                    0
 2, -2305843009213693950, -2305843009213693950
 3, -2305843009213693949, -2305843009213693949

-3,  2305843009213693949,  2305843009213693949
-2,  2305843009213693950,  2305843009213693950
-1,  2305843009213693951,                    0
 0,  2305843009213693952,                    1
 1,  2305843009213693953,                    2


In [3]:
hash(2), hash(2.0)

(2, 2)

In [4]:
hash(1.02)

46116860184273921

In [5]:
type(hash(1.01))

int

In [6]:
hash('a string')

5883023189608143291

In [7]:
hash("A string is an immutable object")

-791094737796391719

### hash 구조의 경우 hash 값이 같을 가능성은 매우 희박하다.

## Hashable object

- Python의 모든 immutable 객체는 hashable 객체이다.
    - bool, int, float, tuple, string, frozenset
- imuutable 객체는 `__hash__()` method를 가지고 있다.
- 객체가 생성되어서 소멸되기 전 까지의 생명주기 동안 객체의 hash 값은 변하지 않는다.

In [8]:
n = 157
print(n.__hash__())
print(hash(n))

157
157


In [9]:
f = 157.157
print(f.__hash__())
print(hash(f))

362017352446574749
362017352446574749


In [11]:
f = 1.0
print(f.__hash__())
print(hash(f))

1
1


In [12]:
b = True
print(b.__hash__())
print(hash(b))

1
1


In [15]:
b = 1.0000000000002
b_ = 1.0000000000001
hash(b), hash(b_)
# 값은 엄청나게 조금 차이나더라도 hash값은 차이가 남.

(461313, 230401)

In [16]:
s = 'string is immuatable.'
print(s.__hash__())
print(hash(s))

7312118434359956082
7312118434359956082


In [17]:
tpl = (1, (4,7,3), ( (1,2,3), (1,2) ))
print(tpl.__hash__())
print(hash(tpl))

-9186792959810590049
-9186792959810590049


# 해쉬불가능(Unhashable) 객체 : list, dictionary, set

In [18]:
t = [1,2]
hash(t)

TypeError: unhashable type: 'list'

In [19]:
d = {'a':1}
hash(d)

TypeError: unhashable type: 'dict'

In [20]:
tpl = (1, (4,7,3), ([1,2,3], (1,2)))
hash(tpl)

TypeError: unhashable type: 'list'

# Dictionary
## Dictionary 문법
- dictionary_name = {key:value, key:value, ⋯..}

## Dictionary 특징
- Dictionary:key-value 쌍 원소의 모음
    - Key : immutable. 키는 중복되지 않는다.
    - value : muatable과 immutable 모두 가능
- 순서가 유지됨.
- Iterable -> 객체의 원소를 한번에 하나씩 반환 가능.
- mutable

### Empty Dictionary 만들기.

In [21]:
empty_dict_1 = {}
empty_dict_2 = dict()
print(empty_dict_1)
print(empty_dict_2)

{}
{}


## Dictionary 만들기.
- Dictionary의 키는 반드시 hashable 객체 (bool, int, float, str, tuple, fronzenset)이여야 한다.

In [24]:
digits = {'one':1, 'two':2, 'three' : 3, 'ten':10}
print(digits)

{'one': 1, 'two': 2, 'three': 3, 'ten': 10}


In [25]:
d1 = dict(enumerate([1,4,10]))
print(d1)

{0: 1, 1: 4, 2: 10}


In [26]:
d2 = dict(zip(['one', 'four', 'ten'], [1,4,10]))
print(d2)

{'one': 1, 'four': 4, 'ten': 10}


In [28]:
d3 = dict((('one',1), ('four', 4), ('ten',10)))
print(d3), type(d3)

{'one': 1, 'four': 4, 'ten': 10}


(None, dict)

In [29]:
d4 = dict([('one',1), ('four',4), ('ten',10)])
print(d4)

{'one': 1, 'four': 4, 'ten': 10}


In [30]:
d5 = dict(one=1, four=4, ten=10)
print(d5)

{'one': 1, 'four': 4, 'ten': 10}


## Key를 이용한 접근

In [31]:
d5['four']

4

In [32]:
d5.get('four')

4

## Key를 이용한 value 수정 / 새 원소 삽입

In [33]:
d5['four'] = 5
print(d5)

{'one': 1, 'four': 5, 'ten': 10}


In [34]:
d5['four'] = 4
print(d5)

{'one': 1, 'four': 4, 'ten': 10}


In [35]:
d5['five'] = 5
print(d5)

{'one': 1, 'four': 4, 'ten': 10, 'five': 5}


## Dictionary 객체

In [36]:
my_dict = {}
print(my_dict.__sizeof__())

48


In [37]:
my_dict['one'] = 1
print(my_dict.__sizeof__())

216


In [39]:
my_dict['two'] = 2
print(my_dict.__sizeof__())

216


In [40]:
my_dict['three'] = 3
print(my_dict.__sizeof__())

216


In [41]:
my_dict['four'] = 4
print(my_dict.__sizeof__())

216


In [42]:
my_dict['five'] = 5
print(my_dict.__sizeof__())

216


In [44]:
my_dict['six'] = 6
print(my_dict)
print(my_dict.__sizeof__())

{'one': 1, 'two': 2, 'three': 3, 'four': 4, 'five': 5, 'six': 6}
344


- dictionary 객체의 크기는 empty일 때 48, 하나가 추가되면 128 byte가 추가된 216이 된다. 원소의 수가 6개가 되면 다시 128이 증가한 344가 된다.
- 여유분을 만들고 그 안에 추가하는 방식으로 딕션너리는 메모리를 관리함.

# Dictionary의 Key는 해쉬가능 객체이어야 한다.

In [45]:
d1 = {1:3, 'a':5, (1,2):5}
print(d1)

{1: 3, 'a': 5, (1, 2): 5}


In [46]:
d2 = {[1,2]:3, 'b':5}

TypeError: unhashable type: 'list'

In [49]:
d1 = {1:3, 'a':5, (1,2):5}
print(hash(1))
print(hash(1.0))
print(d1)

1
1
{1: 3, 'a': 5, (1, 2): 5}


In [50]:
print(d1[1])
print(d1[1.0])

3
3


# 딕셔너리는 키값으로 딕셔너리의 value를 찾는 것이 아닌, 해쉬 값을 통해 value를 찾아온다.

# Dictionary Comprehension

In [52]:
dict_squares = {n:n**2 for n in range(0,10,2)}
dict_squares

{0: 0, 2: 4, 4: 16, 6: 36, 8: 64}

In [53]:
dict_squares = {}

for n in range(0, 10, 2):
    dict_squares[n] = n**2
    
dict_squares

{0: 0, 2: 4, 4: 16, 6: 36, 8: 64}

In [54]:
keys = ['one', 'two', 'three']
d1 = {keys[n]: n+1 for n in range(3)}
d1

{'one': 1, 'two': 2, 'three': 3}

In [55]:
d1.items()

dict_items([('one', 1), ('two', 2), ('three', 3)])

In [57]:
dd = {key:v**2 for (key, v) in d1.items()}
dd

{'one': 1, 'two': 4, 'three': 9}

# Dictionary 연산 

## Pop
- dict.pop(k, [d]): key k를 제거하고 그에 해당하는 value를 반환.
- key가 없는 경우에는, d가 주어지면 d를 반환하고,
- 그렇지 않은 경우에는 KeyError를 발생시킨다.

In [58]:
d1 = {'a': 100, 'b': 25, 'c': 47}
d1.pop('a')

100

In [59]:
d1

{'b': 25, 'c': 47}

In [60]:
d1.pop('who', 4) # d1에는 who가 없어서 4를 반환함.

4

In [61]:
d1

{'b': 25, 'c': 47}

In [62]:
d1.pop('who') #이건 없으니까 오류

KeyError: 'who'

## Popitem
`dict.popitem():`
마지막 key-value 쌍을 제거하고, 제거된 쌍을 튜풀로 반환

In [63]:
d1 = {'a':100, 'b':25, 'c':47}
k, v = d1.popitem()

In [64]:
print(k,v)

c 47


In [65]:
print(d1)

{'a': 100, 'b': 25}


## Fromkeys
`dict.fromkeys(iterable, v=None):`
iterable을 key로 사용하고,
values를 v로 설정한 새 딕셔너리를 생성하여 반환

In [66]:
k = [1,3,5]
d2 = dict.fromkeys(k)
d2

{1: None, 3: None, 5: None}

In [67]:
d3 = dict.fromkeys(k,10)
d3

{1: 10, 3: 10, 5: 10}

In [68]:
d3 = dict.fromkeys('abc', 'ABC')
d3

{'a': 'ABC', 'b': 'ABC', 'c': 'ABC'}

# Copy
- `dict.copy()` : 딕셔너리의 shallow copy 를 반환

In [69]:
d1 = {1:'a', 2:'b', 3:'c'}
d2 = d1
d3 = d1.copy()
d1.popitem()
print(d1)
print(d2)
print(d3)

{1: 'a', 2: 'b'}
{1: 'a', 2: 'b'}
{1: 'a', 2: 'b', 3: 'c'}


# clear
- `dict.clear()` : 딕셔너리의 모든 원소를 제거.

In [70]:
d1.clear()
d1

{}

In [71]:
d2

{}

# get
- `dict.get()` : 딕셔너리에 key가 있느면 해당 value를 반환하고 없으면 default값을 반환.

In [72]:
d1 = {1:'a', 2:'b', 3:'c'}
d1.get(1,'none')

'a'

In [77]:
d1.get(10)  # no outputs

In [76]:
d1.get(10, 'no keys')

'no keys'

# items
- `dict.items()` : 딕셔너리에 있는 모든 (key, values) 쌍을 튜플로 반환

In [79]:
d1

{1: 'a', 2: 'b', 3: 'c'}

In [80]:
d1.items()

dict_items([(1, 'a'), (2, 'b'), (3, 'c')])

# Keys
- `dict.keys()` : 딕셔너리에 있는 모든 key반환

In [81]:
d1

{1: 'a', 2: 'b', 3: 'c'}

In [82]:
d1.keys()

dict_keys([1, 2, 3])

# values
- `dict.values()` : 딕셔너리에 있는 모든 value 반환

In [83]:
d1

{1: 'a', 2: 'b', 3: 'c'}

In [84]:
d1.values()

dict_values(['a', 'b', 'c'])

# Update
`dict.update(iterable)` 
- 딕셔너리에 iterable 삽입. 딕셔너리에 이미 동일한 key가 있는 경우에 해당 value를 새 value로 대치하고, key가 기존 딕셔너리에 없는 경우에 key, value 쌍을 삽입

In [85]:
d1 = {1:'a', 2:'b', 3:'c'}
d2 = {1:'aaa', 4:'bbb'}
d1.update(d2)
d1

{1: 'aaa', 2: 'b', 3: 'c', 4: 'bbb'}

In [86]:
d3 = {'a':'A', 'b':'B', 'c':'C'}
d3.update(c = 5, d=7)
d3

{'a': 'A', 'b': 'B', 'c': 5, 'd': 7}

# Dictionary 관련 메쏘드
- len(object): 컨테이너 내의 항목의 개수 반환
- sum(iterable, start=0): iterable의 합과 start의 합을 반환
- min(iterable): iterable 내의 최솟값을 반환
- max(iterable): iterable 내의 최댓값을 반환
- any(iterable): iterable 내의 어떤 x가 True이면 True를 반환
- all(iterable): iterable 내의 모든 x가 True이면 True를 반환
- sorted(sequence, key=None, reverse=False): sequence의 모든 값을 정렬하여 만든 리스트를 반환
- enumerate(iterable, start=0): start로 시작하는 카운트와 iterable 내의 value의 쌍을 가진 enumerate 객체를 반환
- filter(function, iterable): iterable의 항목 x 중에서 function(x)의 결과가 True인 x를 iterable로 반환

In [88]:
d1 = {1:'a', 2:'b', 3:'c'}
print(len(d1))
print(min(d1))
print(max(d1))
print(sum(d1))
print(sum(d1, 10))
print(sorted(d1))
print(sorted(d1, reverse=True))

3
1
3
6
16
[1, 2, 3]
[3, 2, 1]


In [89]:
d2 = {'a':1, 'b':2, 'c':3}
print(len(d2))
print(min(d2))
print(max(d2))
print(sorted(d2))
print(sorted(d2, reverse=True))

3
a
c
['a', 'b', 'c']
['c', 'b', 'a']


In [95]:
for i, k in enumerate(d2): #앞의 값 a에 0, 1, 2 와 같이 index묶어서 배달.
    print(i, k)

0 a
1 b
2 c


In [92]:
d2

{'a': 1, 'b': 2, 'c': 3}

In [97]:
for s in d2:
    print(s)

a
b
c


In [98]:
for s in d2.keys():
    print(s.upper())

A
B
C


In [99]:
for s in d2.values():
    print(s)

1
2
3
